iree/compiler/Dialect/Flow/Transforms/DeduplicateExecutables.cpp

```diff
@@ -24,111 +25,227 @@ namespace Flow {
+// have the same attributes and same use-def structure.
+//
+// Example:
+//   func @lhs(%arg0 : index) -> index {
+//     %c1 = constant 1 : index
+//     %0 = add %arg0, %c1 : index
+//     return %0 : index
+//   }
+//   func @rhs(%arg0 : index) -> index {
+//     %c1 = constant 1 : index
+//     %0 = add %arg0, %c1 : index
+//     return %0 : index
+//   }
+//
+//   assert(isStructurallyEquivalentTo(lhs.getBody(), rhs.getBody()));
+//
+// TODO(#3996): upstream into mlir::OperationEquivalence if this works.
+// TODO(#3996): add symbol ref comparison (add to BlockAndValueMapping).
+static bool isStructurallyEquivalentTo(Region &lhs, Region &rhs) {
+  BlockAndValueMapping mapping;
+  return isStructurallyEquivalentTo(lhs, rhs, mapping);
+}
+
+static bool isStructurallyEquivalentTo(Region &lhs, Region &rhs,
+                                       BlockAndValueMapping &parentMapping) {
+  // If the region is isolated we don't want to reuse any parent mapping or
+  // pollute it with our mappings.
+  BlockAndValueMapping scopedMapping;
+  BlockAndValueMapping mapping =
+      lhs.isIsolatedFromAbove() ? scopedMapping : parentMapping;
+
+  // Regions can have arguments; ensure they match in type and add the mapping
+  // from lhs -> rhs so that we can track the usage of them within our
+  // descendants.
+  if (lhs.getNumArguments() != rhs.getNumArguments()) return false;
+  for (auto argPair : llvm::zip(lhs.getArguments(), rhs.getArguments())) {
+    auto &lhsArg = std::get<0>(argPair);
+    auto &rhsArg = std::get<1>(argPair);
+    if (lhsArg.getType() != rhsArg.getType()) return false;
+    mapping.map(lhsArg, rhsArg);
   }
 
-  for (auto blockPair : llvm::zip(lhs->getBlocks(), rhs->getBlocks())) {
+  // Walk each block list together. If the sizes differ we'll handle that after
+  // Build the block mapping so that when we recursive into ops we are able to
+  // map lhs->rhs blocks regardless of their order in the region.
+  //
+  // We walk each block list simultaneously and handle length differences after
+  // the loop; this avoids the need for doing an O(n) linked-list size query.
+  if (!compare_ranges(
+          lhs.getBlocks(), rhs.getBlocks(),
+          [&](Block &lhsBlock, Block &rhsBlock) {
+            if (lhsBlock.getNumArguments() != rhsBlock.getNumArguments())
+              return false;
+            for (auto argPair :
+                 llvm::zip(lhsBlock.getArguments(), rhsBlock.getArguments())) {
+              auto &lhsArg = std::get<0>(argPair);
+              auto &rhsArg = std::get<1>(argPair);
+              if (lhsArg.getType() != rhsArg.getType()) return false;
+              mapping.map(lhsArg, rhsArg);
+            }
+            mapping.map(&lhsBlock, &rhsBlock);
+            return true;
+          })) {
+    return false;  // block mismatch
+  }
+
+  // Walk the blocks again now that we have a populated mapping.
+  for (auto blockPair : llvm::zip(lhs.getBlocks(), rhs.getBlocks())) {
```

Comment:
You will lose some precision here if you don't visit domtree parents before domtree children. (so that your mapping is populated for all Value's potentially used in the domtree children, which could be arbitrary op results from any domtree parent block). Note that besides the entry block being first, block list order has nothing to do with domtree order

You can build a domtree and do the walk there, but an easier way is to visit in RPO With [ReversePostOrderTraversal](https://github.com/llvm/llvm-project/blob/93a1fc2e18b452216be70f534da42f7702adbe1d/llvm/include/llvm/ADT/PostOrderIterator.h#L288). (RPO traversal on the CFG ensures that domtree parents are visited before domtree children)

Note that `Block*` implements GraphTraits so ReversePostOrderTraversal should "just work".